### PR TITLE
HPCC-15695 Smart lookup, fix a few bugs + add a few optimizations

### PR DESCRIPTION
**Pull request by jakesmith**
1) OPT: Move broadcast send to others crit around send only
2) OPT: Batch up addRHSRows to avoid some contention
3) OPT: Ensure rightAllocator is unique per channel
4) BUG: Keep rhs rows spillable until after fully prepared HT
Could cause spurious OOM
5) BUG: Protect derived addRHSRow(s) from channel threads
Multiple threads could have both triggered overflowFile
6) BUG: Ensure overflow non-local rows re. pruned
7) BUG: Remove shrink() code from collector::spillRows()
Could cause RoxieMem Invalid pointer on rows
8) BUG: Initialize defaultMaxSpillCost
Could mean spill callbacks were not called

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 1]
```diff
@@ -1057,24 +1058,35 @@ class CInMemJoinBase : public CSlaveActivity, public CAllOrLookupHelper<HELPER>,
...
                 throw MakeActivityException(this, 0, "Out of memory: Unable to add any more rows to RHS");
         }
     }
     void broadcastRHS() // broadcasting local rhs
     {
         Owned<CSendItem> sendItem = broadcaster->newSendItem(bcast_send);
         MemoryBuffer mb;
-        try
+        CThorExpandingRowArray rHSInRowsTemp(*this, sharedRightRowInterfaces);
+        CThorExpandingRowArray pending(*this, sharedRightRowInterfaces);        try
```
Comment (ghalliday) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: typo: missing \n

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 2]
```diff
@@ -1554,21 +1569,15 @@ class CInMemJoinBase : public CSlaveActivity, public CAllOrLookupHelper<HELPER>,
...
     }
-    bool addRHSRow(unsigned slave, const void *row)
+    bool addRHSRows(unsigned slave, CThorExpandingRowArray &inRows, CThorExpandingRowArray &rHSInRowsTemp)
     {
         CThorSpillableRowArray &rows = *rhsSlaveRows.item(slave);
-        return addRHSRow(rows, row);
+        return addRHSRows(rows, inRows, rHSInRowsTemp);
     }
-    virtual bool addRHSRow(CThorSpillableRowArray &rhsRows, const void *row)
+    virtual bool addRHSRows(CThorSpillableRowArray &rhsRows, CThorExpandingRowArray &inRows, CThorExpandingRowArray &rHSInRowsTemp)
```
Comment (ghalliday) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: trivial: rHS is strange capitalisation.  It also doesn't seem to be used

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 3]
```diff
@@ -1938,8 +1948,21 @@ class CLookupJoinActivityBase : public CInMemJoinBase<HTHELPER, IHThorHashJoinAr
...
+        {
+            CThorArrayLockBlock b(rightCollector);
+            if (rightCollector.hasSpilt())
+                return false;
+            /* transfer rows out of collector to perform calc, but we'll keep lock,
+             * so that a request to spill, will block delay, but can still proceed after calculate is done
+             */
+            CThorExpandingRowArray temp(*this);
+            rightCollector.transferRowsOut(temp);
+            uniqueKeys = marker.calculate(rhs, compareRight, false);
```
Comment (ghalliday) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: Shouldn't this be marker.calculate(temp, compareRight, false)??

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 4]
```diff
@@ -2234,18 +2261,23 @@ class CLookupJoinActivityBase : public CInMemJoinBase<HTHELPER, IHThorHashJoinAr
...
             channelDistributor.processDistRight(distChannelStream);
-            stream.setown(channelDistributor.getStream(&rhs));
         }
         catch (IException *e)
         {
             EXCLOG(e, "During channel distribution");
             exception.setown(e);
         }
+
+        /* Now that channel distribution done, remove it's roxiemem memory callback
```
Comment (ghalliday) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: pedantic: no apostrophe

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp (hunk 3 above)
Comment (jakesmith) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: Er, yes!

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: Agree strange capitalization (wasn't new, but I'll change)

It is not used in the based, only in the derived class.

**Comment by ghalliday** [comment]
@jakesmith the code looks good.  Do you know why temp substitution wasn't spotted in testing?

It made it harder to review having lots of changes in the same PR.

@richardkchapman please double check

**Comment by jakesmith** [comment]
>@jakesmith the code looks good. Do you know why temp substitution wasn't spotted in testing?

I would have been by OBT, because it's a 1 slave system, so is in effect performing a SMART,LOCAL.
We should have a SMART,LOCAL test explicitly though. I guess we don't.
I'll add one (via a different JIRA) : https://track.hpccsystems.com/browse/HPCC-15753

**Pull request merged**